akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphZipLatestSpec.scala

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.scaladsl
+
+import akka.actor.ActorSystem
+import akka.stream.testkit.TestPublisher.Probe
+import akka.stream.testkit.scaladsl.{ TestSink, TestSource }
+import akka.stream.{ ActorMaterializer, ClosedShape }
+import akka.testkit.TestKit
+import org.scalacheck.Gen
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.prop.PropertyChecks
+import org.scalatest.{ BeforeAndAfterAll, GivenWhenThen, Matchers, WordSpecLike }
+
+import scala.concurrent.duration._
+import scala.language.postfixOps
+
+class GraphZipLatestSpec
+  extends TestKit(ActorSystem("ZipLatestSpec"))
+  with WordSpecLike
+  with Matchers
+  with BeforeAndAfterAll
+  with PropertyChecks
+  with GivenWhenThen
+  with ScalaFutures {
+  implicit val materializer = ActorMaterializer()
+  override def afterAll = TestKit.shutdownActorSystem(system)
+  implicit val patience = PatienceConfig(5 seconds)
+
+  "ZipLatest" must {
+    "only emit when at least one pair is available" in {
+      val (probe, bools, ints) = testGraph[Boolean, Int]
+
+      Given("request for one element")
+      probe.request(1)
+
+      And("an element pushed on one of the sources")
+      bools.sendNext(true)
+
+      Then("does not emit yet")
+      probe.expectNoMessage(0 seconds)
+
+      And("an element pushed on the other source")
+      ints.sendNext(1)
+
+      Then("emits a single pair")
+      probe.expectNext((true, 1))
+    }
+
+    "emits as soon as one source is available" in {
+      val (probe, bools, ints) = testGraph[Boolean, Int]
+
+      Given("request for 3 elements")
+      probe.request(3)
+
+      And("a first element pushed on either source")
+      bools.sendNext(true)
+      ints.sendNext(1)
+
+      And("then 2 elements pushed only on one source")
+      ints.sendNext(2)
+      ints.sendNext(3)
+
+      Then("3 elements are emitted")
+      probe.expectNext((true, 1))
+      probe.expectNext((true, 2))
+      probe.expectNext((true, 3))
+    }
+
+    "does not emit the same pair upon two pulls with value types" in {
```

Comment:
this is what the equality comparison is for: we don't want to re-emit the same output when pulling again on the source. But we don't dedup when new values are pushed from upstream.

Reply:
The question here is what the "same" is. I argue it is the same instance of tuple, not a tuple that is equal to a previously created tuple. Since it shouldn't be about equality I think we can get away with tracking the fact that a value came in and we haven't emitted anything yet. (Potentially much cheeper than keeping the last emitted element in memory)

Reply:
yes you are right. It would also be clearer. I'll see about modifying this, hopefully this afternoon

Reply:
Great, thanks!